pkg/resources/password_policy.go

```diff
@@ -0,0 +1,447 @@
+	}
+
+	input := &snowflake.PasswordPolicyReadInput{
+		Database:   d.Get("database").(string),
+		Schema:     d.Get("schema").(string),
+		ObjectName: d.Get("name").(string),
+	}
+
+	stmt, err := manager.Read(input)
+	if err != nil {
+		return fmt.Errorf("couldn't generate describe statement: %w", err)
+	}
+
+	db := meta.(*sql.DB)
+	rows, err := db.Query(stmt)
+	if err != nil {
+		return fmt.Errorf("error querying password policy: %w", err)
+	}
+
+	defer rows.Close()
+	output, err := manager.Parse(rows)
+	if err != nil {
+		return fmt.Errorf("failed to parse result of describe: %w", err)
+	}
+
+	if err = d.Set("min_length", output.MinLength); err != nil {
+		return fmt.Errorf("error setting min_length: %w", err)
+	}
+	if err = d.Set("max_length", output.MaxLength); err != nil {
+		return fmt.Errorf("error setting max_length: %w", err)
+	}
+	if err = d.Set("min_upper_case_chars", output.MinUpperCaseChars); err != nil {
+		return fmt.Errorf("error setting min_upper_case_chars: %w", err)
+	}
+	if err = d.Set("min_lower_case_chars", output.MinLowerCaseChars); err != nil {
+		return fmt.Errorf("error setting min_lower_case_chars: %w", err)
+	}
+	if err = d.Set("min_numeric_chars", output.MinNumericChars); err != nil {
+		return fmt.Errorf("error setting min_numeric_chars: %w", err)
+	}
+	if err = d.Set("min_special_chars", output.MinSpecialChars); err != nil {
+		return fmt.Errorf("error setting min_special_chars: %w", err)
+	}
+	if err = d.Set("max_age_days", output.MaxAgeDays); err != nil {
+		return fmt.Errorf("error setting max_age_days: %w", err)
+	}
+	if err = d.Set("max_retries", output.MaxRetries); err != nil {
+		return fmt.Errorf("error setting max_retries: %w", err)
+	}
+	if err = d.Set("lockout_time_mins", output.LockoutTimeMins); err != nil {
+		return fmt.Errorf("error setting lockout_time_mins: %w", err)
+	}
+
+	return nil
+}
+
+// UpdatePasswordPolicy implements schema.UpdateFunc.
+func UpdatePasswordPolicy(d *schema.ResourceData, meta interface{}) error {
+	manager, err := snowflake.NewPasswordPolicyManager()
+	if err != nil {
+		return fmt.Errorf("couldn't create password policy builder: %w", err)
+	}
+
+	runAlter := false
+	alterInput := &snowflake.PasswordPolicyUpdateInput{
+		PasswordPolicy: snowflake.PasswordPolicy{
+			SchemaObjectIdentifier: snowflake.SchemaObjectIdentifier{
+				Database:   d.Get("database").(string),
+				Schema:     d.Get("schema").(string),
+				ObjectName: d.Get("name").(string),
+			},
+		},
+	}
+	runUnset := false
+	unsetInput := &snowflake.PasswordPolicyUpdateInput{
+		PasswordPolicy: snowflake.PasswordPolicy{
+			SchemaObjectIdentifier: snowflake.SchemaObjectIdentifier{
+				Database:   d.Get("database").(string),
+				Schema:     d.Get("schema").(string),
+				ObjectName: d.Get("name").(string),
+			},
+		},
+	}
+
+	if d.HasChange("min_length") {
+		val, ok := d.GetOk("min_length")
+		if ok {
+			alterInput.MinLength = val.(int)
+			alterInput.MinLengthOk = true
+			runAlter = true
```

Comment:
is this flag really necessary? perhaps it could be set internally? Also how to differentiate between a non-set value, and a set value? ex, lets say MinLength is not set. It would default to false when you instantiate the struct. So it might look like its being set to false when actually its not being set at all.

Reply:
1. yes, an update is an ALTER SET query and/or an ALTER UNSET query, we need to know which ones are necessary.
2. That's what the ok flag is for for each field (true = flag is present)